p2p/listener.go

```diff
@@ -27,25 +32,21 @@ type ListenerRegistry struct {
 	lk        sync.Mutex
 }
 
-func (r *ListenerRegistry) lock(l Listener) error {
+// Register registers listenerInfo into this registry and starts it
+func (r *ListenerRegistry) Register(l Listener) error {
 	r.lk.Lock()
+	defer r.lk.Unlock()
 
 	if _, ok := r.Listeners[getListenerKey(l)]; ok {
-		r.lk.Unlock()
 		return errors.New("listener already registered")
 	}
-	return nil
-}
 
-func (r *ListenerRegistry) unlock() {
-	r.lk.Unlock()
-}
-
-// Register registers listenerInfo in this registry
-func (r *ListenerRegistry) Register(l Listener) {
-	defer r.lk.Unlock()
+	if err := l.start(); err != nil {
```

Comment:
You're probably going to want to:

1. Lock
2. Register
3. **unlock**
4. Start
5. unregister (under a lock) on error.

That is, starting under a lock is probably not going to work out well.